pkg/reconciler/v1alpha1/eventlistener/eventlistener.go

```diff
@@ -81,134 +83,132 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+		// Create the EventListener Service
+		_, err = c.KubeClientSet.CoreV1().Services(service.Namespace).Create(service)
 		if err != nil {
-			c.Logger.Errorf("Error updating EventListener Deployment: %s", err)
+			c.Logger.Errorf("Error creating EventListener Service: %s", err)
+			el.Status.SetStatus(v1alpha1.ServiceConditionType, err)
 			return err
 		}
-		c.Logger.Info("Updated EventListener Deployment %s in Namespace %s", el.Name, el.Namespace)
+		c.Logger.Infof("Created EventListener Service %s in Namespace %s", service.Name, service.Namespace)
+		el.Status.SetStatus(v1alpha1.ServiceConditionType, nil)
+		el.Status.Configuration.Hostname = ListenerHostname(service.Name, service.Namespace)
 	}
 
-	// TODO: This is an example Service that we will probably want to modify in the future
-	service := &corev1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			// Create the EventListener's Service in the same Namespace as where the
-			// EventListener was created
-			Namespace: el.Namespace,
-			// Give the Service the same name as the EventListener
-			Name: el.Name,
-			// If our EventListener is deleted, then its Service should be as well
-			OwnerReferences: []metav1.OwnerReference{*el.GetOwnerReference()},
-			Labels:          labels,
-		},
-		Spec: corev1.ServiceSpec{
-			Selector: map[string]string{"app": el.Name},
-			Type:     corev1.ServiceTypeLoadBalancer,
-			Ports: []corev1.ServicePort{
-				{
-					Protocol: corev1.ProtocolTCP,
-					Port:     int32(Port),
+	if _, err := c.KubeClientSet.AppsV1().Deployments(el.Namespace).Get(el.Name, metav1.GetOptions{}); err != nil {
+		// Create the EventListener Deployment
+		deployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				// Create the EventListener's Deployment in the same Namespace as where
+				// the EventListener was created
+				Namespace: el.Namespace,
+				// Give the Deployment the same name as the EventListener
+				Name: el.Name,
+				// If our EventListener is deleted, then its Deployment should be as well
+				OwnerReferences: []metav1.OwnerReference{*el.GetOwnerReference()},
+				Labels:          labels,
+			},
+			Spec: appsv1.DeploymentSpec{
+				Selector: &metav1.LabelSelector{
+					MatchLabels: labels,
+				},
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: labels,
+					},
+					Spec: corev1.PodSpec{
+						// TODO: Do not make the assumption that the EventListener/Deployment is in the tekton-pipelines namespace
+						Containers: []corev1.Container{
+							{
+								Name:  "event-listener",
+								Image: *elImage,
+								Ports: []corev1.ContainerPort{
+									{
+										ContainerPort: int32(Port),
+									},
+								},
+								Env: []corev1.EnvVar{
+									corev1.EnvVar{
+										Name:  "LISTENER_NAME",
+										Value: el.Name,
+									},
+									corev1.EnvVar{
+										Name:  "LISTENER_NAMESPACE",
+										Value: el.Namespace,
+									},
+								},
+							},
+						},
+					},
```

Comment:
it looks like we have some opportunities to set a good precedent early here and break this into separate functions that can be unit tested, e.g.:

- A function to create the needed service, params would include the create function (`c.KubeClientSet.CoreV1().Services(service.Namespace).Create` (passing in functions like this makes unit testing way easier - that way you don't need tho whole reconciler and the unit test can just pass in a function that does whatever you like, in this case either succeeds or has an error)
- A similar function to create the Deployment
- Maybe even a function to get the status?